multitenant: add AdminRelocateRange capability

ecwall opened the pull request:
Fixes #98467

1) Add a new `tenantcapabilities.AdminRelocateRange` capability.
2) Check capability in `Authorizer.HasCapabilityForBatch` for
   `AdminChangeReplicas`, `AdminRelocateRange`, `AdminTransferLease`.
3) Remove unused `SkipSQLSystemTentantCheck`.

Release note: None

ecwall committed:
multitenant: reduce capability unrelated test churn

Exclude capabilities from `tenant_builtins` logic test. These tests do not set
or depend on capabilities and they are tested elsewhere.

Release note: None

ecwall committed:
multitenant: reduce capability unrelated test churn

Change `watcher_test` to only output capabilities that were altered for the
test.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/98483)
<!-- Reviewable:end -->

knz reviewed (commented):
Don't forget to update the tests reporting failing in CI.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 7 of 7 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQQ528N-5lkjWvankRM:b-hgu75z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4Tw30v9G69XkBoJx:-NQQ4Tw30v9G69XkBoJy:bmffjds) ([raw file](https://github.com/cockroachdb/cockroach/blob/779b7efc8e4ee1010aa77a4fc2049ca86737fc1e/pkg/multitenant/tenantcapabilities/capabilities.go#L101)):*
> ```Go
> //
> //	https://github.com/cockroachdb/cockroach/issues/96736.
> var DefaultCapabilities TenantCapabilities
> ```

tip: make this a function (`func DefaultCapabilities() (res TenantCapabilities) { return res }` to ensure that no code can modify this global variable by accident. This is useful because Go doesn't have the `const` keyword.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4jgp7fDwVhmHhx9M:-NQQ4jgq9TXJ0YH8jWhi:b-2y506o) ([raw file](https://github.com/cockroachdb/cockroach/blob/779b7efc8e4ee1010aa77a4fc2049ca86737fc1e/pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go#L20)):*
> ```Go
> 
> func init() {
> 	tenantcapabilities.DefaultCapabilities = &TenantCapabilities{}
> ```

Why do you need this?
___
*[`pkg/sql/opt_exec_factory.go` line 2041 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4xgO68lfZ3os-Cd9:-NQQ4xgPBlZmXOue-vGk:bsierfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1b0ceb11a4a8384b8f6866ec0663fc16ffdf92/pkg/sql/opt_exec_factory.go#L2041)):*
> ```Go
> 	fromStoreID tree.TypedExpr,
> ) (exec.Node, error) {
> 	if relocateSubject == tree.RelocateLease {
> ```

Do you expect we'll use separate capabilities to relocate replicas and leases? Why?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4Tw30v9G69XkBoJx:-NQQ5FrD5HYwxGy7nRMO:bs1s2gv) ([raw file](https://github.com/cockroachdb/cockroach/blob/779b7efc8e4ee1010aa77a4fc2049ca86737fc1e/pkg/multitenant/tenantcapabilities/capabilities.go#L101)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

tip: make this a function (`func DefaultCapabilities() (res TenantCapabilities) { return res }` to ensure that no code can modify this global variable by accident. This is useful because Go doesn't have the `const` keyword.
</blockquote></details>

Sorry I forgot this is an interface. So this instead:
```go

func DefaultCapabilities{} TenantCapabilities { return defaultCaps }
var defaultCaps TenantCapabilities

// RegisterDefaultCapabilities will be called from the pb package.
func RegisterDefaultCapabilities(caps TenantCapabilities) { defaultCaps = caps }
```


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ5ayn0eO44-29RZe9:-NQQ5ayn0eO44-29RZeA:b-22gcgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/779b7efc8e4ee1010aa77a4fc2049ca86737fc1e/pkg/multitenant/tenantcapabilities/capabilities.go#L109)):*
> ```Go
> 	var builder strings.Builder
> 	builder.WriteByte('{')
> 	isFirst := true
> ```

tip to remove this `isFirst` thing

```go
space := ""
builder.WriteByte('{'}
for _, capID := range ... {
   ...
   if value !=  defaultValue {
       builder.WriteString(space)
       ...
       space = " "
   }
}
```


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQQA4Ij2JUfEz_MOivS:b-bw8e14)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 101 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4Tw30v9G69XkBoJx:-NQQAFAc5xgo6bePN1sA:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/779b7efc8e4ee1010aa77a4fc2049ca86737fc1e/pkg/multitenant/tenantcapabilities/capabilities.go#L101)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Sorry I forgot this is an interface. So this instead:
```go

func DefaultCapabilities{} TenantCapabilities { return defaultCaps }
var defaultCaps TenantCapabilities

// RegisterDefaultCapabilities will be called from the pb package.
func RegisterDefaultCapabilities(caps TenantCapabilities) { defaultCaps = caps }
```
</blockquote></details>

Updated.
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 109 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ5ayn0eO44-29RZe9:-NQQDVRTDSP70NvnwHFZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/779b7efc8e4ee1010aa77a4fc2049ca86737fc1e/pkg/multitenant/tenantcapabilities/capabilities.go#L109)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

tip to remove this `isFirst` thing

```go
space := ""
builder.WriteByte('{'}
for _, capID := range ... {
   ...
   if value !=  defaultValue {
       builder.WriteString(space)
       ...
       space = " "
   }
}
```
</blockquote></details>

Done.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4jgp7fDwVhmHhx9M:-NQQAOOu2j7EfWMsUnZy:bithjfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/779b7efc8e4ee1010aa77a4fc2049ca86737fc1e/pkg/multitenant/tenantcapabilities/tenantcapabilitiespb/capabilities.go#L20)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Why do you need this?
</blockquote></details>

`tenantcapabilitiespb` depends on `tenantcapabilities` so this is to prevent a dependency cycle.

I made this change ATM to reduce test churn, but this code will be revisited when I work on actually supporting default capability values in https://github.com/cockroachdb/cockroach/issues/96736.
___
*[`pkg/sql/opt_exec_factory.go` line 2041 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4xgO68lfZ3os-Cd9:-NQQDbW678AXYx-Vb14F:b-nutz94) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1b0ceb11a4a8384b8f6866ec0663fc16ffdf92/pkg/sql/opt_exec_factory.go#L2041)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Do you expect we'll use separate capabilities to relocate replicas and leases? Why?
</blockquote></details>

They have different endpoints. Isn't there a capability per endpoint or is this an exception?


<!-- Sent from Reviewable.io -->

ecwall committed:
multitenant: add AdminRelocateRange capability

Fixes #98467

1) Add a new `tenantcapabilities.AdminRelocateRange` capability.
2) Check capability in `Authorizer.HasCapabilityForBatch` for
   `AdminChangeReplicas`, `AdminRelocateRange`, `AdminTransferLease`.
3) Remove unused `SkipSQLSystemTentantCheck`.

Release note: None

knz reviewed (commented):
> Reviewed 5 of 16 files at r5, 4 of 7 files at r7, 9 of 9 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/sql/opt_exec_factory.go` line 2041 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4xgO68lfZ3os-Cd9:-NQQGI337MDtWSdwbUwP:b-8q69n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1b0ceb11a4a8384b8f6866ec0663fc16ffdf92/pkg/sql/opt_exec_factory.go#L2041)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

They have different endpoints. Isn't there a capability per endpoint or is this an exception?
</blockquote></details>

IMHO capabilities should be per "user-facing feature". In this case, ALTER RANGE RELOCATE is just "one feature", wether you move replicas or leases.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
As discussed: it may be valuable to group the various Relocate operations under a single capability to simplify operational overhead.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQQPH7JArfzOkouWN4S:b9f2eg9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)



<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/sql/opt_exec_factory.go` line 2041 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQ4xgO68lfZ3os-Cd9:-NQQHPYx0gVkH6VyZiKH:b-xgptje) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1b0ceb11a4a8384b8f6866ec0663fc16ffdf92/pkg/sql/opt_exec_factory.go#L2041)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

IMHO capabilities should be per "user-facing feature". In this case, ALTER RANGE RELOCATE is just "one feature", wether you move replicas or leases.
</blockquote></details>

I changed all of these to use one `can_admin_relocate_range` capability: 
```
RELOCATE LEASE -> AdminTransferLease
RELOCATE VOTERS -> AdminChangeReplicas
RELOCATE NONVOTERS -> AdminChangeReplicas
EXPERIMENTAL_RELOCATE LEASE -> AdminTransferLease
EXPERIMENTAL_RELOCATE VOTERS -> AdminRelocateRange
EXPERIMENTAL_RELOCATE NONVOTERS -> AdminRelocateRange
```


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 10 of 10 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQZktN6WN8lpVH553f:-NQQZktN6WN8lpVH553g:b-o5navo) ([raw file](https://github.com/cockroachdb/cockroach/blob/dab98eeec31beeec8d5dccee13228fd62e051f56/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L133)):*
> ```Go
> 
> 	// TODO(ecwall): The following should also be authorized via specific capabilities.
> 	kvpb.AdminMerge: noCapCheckNeeded,
> ```

NB: it's absolutely essential we either restrict AdminMerge to a capability or change this to `onlySystemTenant`. Having secondary tenants use AdminMerge directly without verification is dangerous.

(You can do this in this PR or another followup PR)


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQZktN6WN8lpVH553f:-NQQaaCl1_Wdld5tb2Nt:b38qrrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dab98eeec31beeec8d5dccee13228fd62e051f56/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L133)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

NB: it's absolutely essential we either restrict AdminMerge to a capability or change this to `onlySystemTenant`. Having secondary tenants use AdminMerge directly without verification is dangerous.

(You can do this in this PR or another followup PR)
</blockquote></details>

I'll do this in a follow up because I have to familiarize myself with `AdminMerge` first. I am not sure I have a test case for it in `multitenant_admin_function_test.go`.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 133 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQQZktN6WN8lpVH553f:-NQQeAY06AJ8TLNecsMJ:bf9l4oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/dab98eeec31beeec8d5dccee13228fd62e051f56/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L133)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

I'll do this in a follow up because I have to familiarize myself with `AdminMerge` first. I am not sure I have a test case for it in `multitenant_admin_function_test.go`.
</blockquote></details>

You cannot trigger it from SQL so that is why you didn't see it before. But it can be used by end-users via the SendKVBatch HTTP API.


<!-- Sent from Reviewable.io -->

ecwall commented:
bors r=knz

yuzefovich commented:
Random drive-by comment: I see that CI failed on `TestRelocateVoters` on this PR and seems related, is this an old flake or a new flake being introduced by this PR?

ecwall commented:
bors r-

ecwall commented:
> Random drive-by comment: I see that CI failed on `TestRelocateVoters` on this PR and seems related, is this an old flake or a new flake being introduced by this PR?

I think it may be some non-deterministic behavior that needs to be accounted for in this PR. I will update the tests to be more lenient.

ecwall committed:
multitenant: deflake multitenant_admin_function_test

Ignore ranges that can change between test runs in result verification.

Release note: None

arulajmani reviewed (dismissed):
Mostly looks good. Few minor comments that we should address before merging, however. 

> Reviewed 1 of 1 files at r1, 15 of 15 files at r15, 1 of 1 files at r16, 7 of 7 files at r17, 11 of 12 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQRh02VEra0AjiIEIle:b-h4i8vp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 169 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQReowJ80Sp9zfHbG7Q:-NQReowJ80Sp9zfHbG7R:b-atzxzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f8cbe82937832ded6f616ef8a74726edc94781/pkg/multitenant/tenantcapabilities/capabilities.go#L169)):*
> ```Go
> 	// because excessive KV range scatter can overwhelm the storage
> 	// cluster.
> 	CanAdminScatter CapabilityID = iota + MinCapabilityID // can_admin_scatter
> ```

This doesn't feel idiomatic. Can we please revert this change? 
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 99 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRf60aCdZm5xgDNCab:-NQRf60aCdZm5xgDNCac:bk19qli) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L99)):*
> ```Go
> // defaultCaps is the default state of capabilities.
> // TODO(ewall): Redesign as part of
> //
> ```

nit: remove this new line. 
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 100 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfBIe9LxyRou4o0hA:-NQRfBIe9LxyRou4o0hB:bp3b70q) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L100)):*
> ```Go
> // TODO(ewall): Redesign as part of
> //
> //	https://github.com/cockroachdb/cockroach/issues/96736.
> ```

Let's create a new issue for this with a proposed redesign. This is mostly for me, and other readers, who would want to follow along with what you may have in mind when you say redesign. If you don't want to create an issue here, you could consider adding more words to the TODO with what you have in mind. 
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 109 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfYe80JCJs3iHMkUN:-NQRfYe80JCJs3iHMkUO:bgwln9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L109)):*
> ```Go
> // longer match DefaultCapabilities. This is different from
> // TenantCapabilities.String which only prints non-zero value fields.
> func AlteredCapabilitiesString(capabilities TenantCapabilities) string {
> ```

Let's make this a testing only function and put it in the `testutils` package instead.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 75 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfvvP-dxCbbw6YBzn:-NQRfvvP-dxCbbw6YBzo:b-6ebhz3) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L75)):*
> ```Go
> 			// CanAdminSplit and CanAdminScatter have special handling here for tests
> 			// that rely on them defaulting to true.
> 			// TODO(ewall): Redesign as part of
> ```

Again, add more words about what you mean here by redesign. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/rangefeed_errors` line 93 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRg4xsANfxI16yBl2_:-NQRg4xsANfxI16yBl2a:bpa4b2w) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/rangefeed_errors#L93)):*
> ```
> ten=11 altered-cap={}
> ten=12 altered-cap={can_admin_split:true}
> ten=50 altered-cap={}
> ```

Instead of printing empty capabilities, let's print something like `{default}`. Look at `PrintSpanConfigDiffedAgainstDefaults` and how it's used in datadriven tests for inspiration. 
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher_test.go` line 228 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRgDMb1aXfHD3T5FMb:-NQRgDMb1aXfHD3T5FMc:b-sdqtwq) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher_test.go#L228)):*
> ```Go
> 				lastUpdateTS = ts.Clock().Now()
> 
> 			case "get-altered-capabilities":
> ```

Let's continue calling the directive `get-capabilities`. However, the fact that we're diffing against default, should be highlighted in the DSL description in the comment. 


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 169 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQReowJ80Sp9zfHbG7Q:-NQRiC3t5peLfc-ZrPRJ:b-ey4wg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f8cbe82937832ded6f616ef8a74726edc94781/pkg/multitenant/tenantcapabilities/capabilities.go#L169)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This doesn't feel idiomatic. Can we please revert this change? 
</blockquote></details>

Previously the hard coded `1` value being passed to `StringToEnumValueMap` and `EnumValue` represented the number of `_`'s at the start of this const block.

Using `MinCapabilityID` in both places shows this relationship whereas the previous code did not indicate one at all. 

Do you have an idiomatic way of doing this?
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 99 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRf60aCdZm5xgDNCab:-NQRpOoA0-zI3nNaZy6t:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L99)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: remove this new line. 
</blockquote></details>

Removed.
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 100 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfBIe9LxyRou4o0hA:-NQRrWJgEUdNMyHg4TE6:b-1ohk46) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L100)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's create a new issue for this with a proposed redesign. This is mostly for me, and other readers, who would want to follow along with what you may have in mind when you say redesign. If you don't want to create an issue here, you could consider adding more words to the TODO with what you have in mind. 
</blockquote></details>

I am no longer sure if it will change so I will remove that TODO for now.
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 109 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfYe80JCJs3iHMkUN:-NQRv5Fn8mB2Y3CsxWSV:b6pcwwu) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L109)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's make this a testing only function and put it in the `testutils` package instead.
</blockquote></details>

This creates a circular dependency so I will leave it here.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go` line 75 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfvvP-dxCbbw6YBzn:-NQRw59i9iQS0uH_cA51:b-fijhhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/tenantcapabilitiesauthorizer/authorizer.go#L75)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Again, add more words about what you mean here by redesign. 
</blockquote></details>

Removing this comment.
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/rangefeed_errors` line 93 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRg4xsANfxI16yBl2_:-NQRwAN1-XL5c9mfKNXO:b-6r01tg) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/testdata/rangefeed_errors#L93)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Instead of printing empty capabilities, let's print something like `{default}`. Look at `PrintSpanConfigDiffedAgainstDefaults` and how it's used in datadriven tests for inspiration. 
</blockquote></details>

Switched to `{default}`
___
*[`pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher_test.go` line 228 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRgDMb1aXfHD3T5FMb:-NQRwCn9CjlfQhjKV0_F:b-t3oyoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher/watcher_test.go#L228)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Let's continue calling the directive `get-capabilities`. However, the fact that we're diffing against default, should be highlighted in the DSL description in the comment. 
</blockquote></details>

Renamed back.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
While working on https://github.com/cockroachdb/cockroach/pull/98546, I realized that we're not adding E2E tests in `/pkg/ccl/multitenantccl/tenantcapabilitiesccl` for this functionality. Could you add tests for these capabilities in there as well? Separately, it would be nice to go back and add tests for other capabilities there too -- I added some for `can_admin_unsplit`, but we're still missing`disable_admin_scatter` for now. 

Also, could I request you to hold off on merging this until I land that release blocker? That PR moves a bunch of things around and changes the `Authorizer` tests in a way that might cause merge conflicts. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQSNm617za_jhgYMCW-:bqksdk8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 109 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfYe80JCJs3iHMkUN:-NQSN8FL662oVi8fV2uw:b-72abrt) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L109)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

This creates a circular dependency so I will leave it here.
</blockquote></details>

To be clear, I meant the `tenantcapabilitiestestutils` package. IIUC, it shouldn't create a circular dependency as we already import the `tenantcapabilities` package in there. Is that the package you had in mind as well? 


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 109 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQRfYe80JCJs3iHMkUN:-NQSg_GA8BsxgcAZrj83:b-oibfwa) ([raw file](https://github.com/cockroachdb/cockroach/blob/43b61f79b4e1f6a3c12c854a08757647e58d7d39/pkg/multitenant/tenantcapabilities/capabilities.go#L109)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

To be clear, I meant the `tenantcapabilitiestestutils` package. IIUC, it shouldn't create a circular dependency as we already import the `tenantcapabilities` package in there. Is that the package you had in mind as well? 
</blockquote></details>

Here is the cycle:
```
package github.com/cockroachdb/cockroach/pkg/multitenant/tenantcapabilities/tenantcapabilitieswatcher
	imports github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed
	imports github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord
	imports github.com/cockroachdb/cockroach/pkg/gossip
	imports github.com/cockroachdb/cockroach/pkg/rpc
	imports github.com/cockroachdb/cockroach/pkg/multitenant/tenantcapabilities
	imports github.com/cockroachdb/cockroach/pkg/multitenant/tenantcapabilities/tenantcapabilitiestestutils
	imports github.com/cockroachdb/cockroach/pkg/multitenant/tenantcapabilities: import cycle not allowed
```

I created this https://github.com/cockroachdb/cockroach/issues/98548 to look at it separately since this PR is not trying to solve that.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
End to end tests already exist in `multitenant_admin_function_test`. `tenantcapabilitiesccl` does not support the cluster setup required for these tests.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQVHBrrBeNmDToYM-mb:b3ptoiz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
Thanks the structure of this PR looks fine to me now. I have a suggestion to clarify the capability ID numbering, see below.

> Reviewed 5 of 16 files at r14, 4 of 6 files at r29, 11 of 12 files at r30, 1 of 1 files at r31.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQVY3wG8doV7gjYuusA:b-xx8l6i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 169 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQReowJ80Sp9zfHbG7Q:-NQVXCgE1DB4yvcr4a-D:b-azbwx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8f8cbe82937832ded6f616ef8a74726edc94781/pkg/multitenant/tenantcapabilities/capabilities.go#L169)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Previously the hard coded `1` value being passed to `StringToEnumValueMap` and `EnumValue` represented the number of `_`'s at the start of this const block.

Using `MinCapabilityID` in both places shows this relationship whereas the previous code did not indicate one at all. 

Do you have an idiomatic way of doing this?
</blockquote></details>

This looks fine to me now.
___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 223 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQVXQDK0BHFiTBGTqd3:-NQVXQDK0BHFiTBGTqd4:b-e7gqg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f099e40eb4e50bcf50a4ab79b0a6607a1764f7c/pkg/multitenant/tenantcapabilities/capabilities.go#L223)):*
> ```Go
> 
> 	// MaxCapabilityID is the value of the maximum CapabilityID.
> 	MaxCapabilityID CapabilityID = iota + MinCapabilityID - 1
> ```

You could call this constant just `NumCapabilityIDs` (without an `=` assignment)

Then use `NumCapabilityIDs-MinCapabilityID` in various places below. 


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
Thanks

> Reviewed 16 of 16 files at r32, 2 of 2 files at r33, 6 of 6 files at r34, 12 of 12 files at r35, 1 of 1 files at r36, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQZMty1-svII8e_99JX:b-zcjvte)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)



<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 223 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQVXQDK0BHFiTBGTqd3:-NQ_Bpyy257JFPChE_Pv:b18mmsa) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f099e40eb4e50bcf50a4ab79b0a6607a1764f7c/pkg/multitenant/tenantcapabilities/capabilities.go#L223)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

You could call this constant just `NumCapabilityIDs` (without an `=` assignment)

Then use `NumCapabilityIDs-MinCapabilityID` in various places below. 
</blockquote></details>

I think that only works if `MinCapabilityID` is 0. I was trying to come up with a general purpose way of having `MinCapabilityID` point to the lowest value and `MaxCapabilityID` point to the highest without requiring manual changes when capabilities are added or removed.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 223 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQVXQDK0BHFiTBGTqd3:-NQ_LMLt4_3Iu7jzN6kR:b-1sd25o) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f099e40eb4e50bcf50a4ab79b0a6607a1764f7c/pkg/multitenant/tenantcapabilities/capabilities.go#L223)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

I think that only works if `MinCapabilityID` is 0. I was trying to come up with a general purpose way of having `MinCapabilityID` point to the lowest value and `MaxCapabilityID` point to the highest without requiring manual changes when capabilities are added or removed.
</blockquote></details>

With all due respect, this complexity here (and the risks associated with having people thinking about when to use `+ 1` vs `- 1`) is a lot of complexity just to enable this `stringToCapabilityIDMap` below. This is not even particularly "interesting" because the real solution would be to have `stringer` generate the mapping for us - so we wouldn't even need to implement `CapabilityIDFromString` ourselves.

I recommend you find a way forward that clarifies this and removes footguns, and see if one of the possible solutions would be to hand-code the full map for now and also make a plan to customize `stringer` in the future.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
@ecwall I checked out your PR, and seems like this is the source of your dependency cycle:

https://github.com/cockroachdb/cockroach/blob/de3a5e5d6e07ae00b4a2ce8b5261cd7057fc91ad/pkg/multitenant/tenantcapabilities/capabilities.go#L131-L133

This feels off -- this is production code, but we're calling into a testing only function to format an `Entry`. I don't see why a commit that's trying to reduce test churn is also changing how the struct is printed elsewhere, such as in logs. 

If you wanted to get rid of this dependency, you could write a `Print` function for the `Entry` struct in `tenantcapabilitiestestutils` that formats things using this diffing method like you want _just_ for tests. 

All this being said, if @knz is okay with the way this code looks, I'll step aside. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQ_UYB460FWfBEs4qAy:bnrpyes)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)



<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
I moved this function to `tenantcapabilitiestestutils.AlteredCapabilitiesString` and modified the test code to customize the output instead of relying on `Update.String()`, `Entry.String()`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-:-NQ_q33w5QJFWTB62Zxk:b-xdaul8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @knz, and @rafiss)

___
*[`pkg/multitenant/tenantcapabilities/capabilities.go` line 223 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/98483#-NQVXQDK0BHFiTBGTqd3:-NQ_pEbW0UYAeDNkndz_:bl18447) ([raw file](https://github.com/cockroachdb/cockroach/blob/2f099e40eb4e50bcf50a4ab79b0a6607a1764f7c/pkg/multitenant/tenantcapabilities/capabilities.go#L223)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

With all due respect, this complexity here (and the risks associated with having people thinking about when to use `+ 1` vs `- 1`) is a lot of complexity just to enable this `stringToCapabilityIDMap` below. This is not even particularly "interesting" because the real solution would be to have `stringer` generate the mapping for us - so we wouldn't even need to implement `CapabilityIDFromString` ourselves.

I recommend you find a way forward that clarifies this and removes footguns, and see if one of the possible solutions would be to hand-code the full map for now and also make a plan to customize `stringer` in the future.
</blockquote></details>

Talked about this offline - I will revert this for now and look at using a code generator to make the map and slice instead of using stringer + these min/max constants.

Though stringer already exists, it seems to be the wrong tool to use here if this calculation of min/max is too convoluted.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
> Reviewed 2 of 14 files at r41, 2 of 12 files at r42, 1 of 1 files at r43.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/98483)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @ecwall, and @rafiss)



<!-- Sent from Reviewable.io -->

arulajmani commented:
> I moved this function to tenantcapabilitiestestutils.AlteredCapabilitiesString and modified the test code to customize the output instead of relying on Update.String(), Entry.String().

Thanks for making the change.

ecwall commented:
bors r=knz

knz commented:
the function TestExperimentalRelocateNonVoters flaked in CI. If we're not 100% sure about it, i'd rather have it skipped before it causes CI failures for others.
